### DISCUSSION TITLE
PPP

### DISCUSSION THREAD

**Discussion by Hsg2311** [MyeongWoonJang/cpp-study#171] [General]
## PPP
PPP는 Pseudocode Programming Process의 약자입니다.
영어 단어를 봤을 때, 대충 감이 오시죠? 맞습니다! 의사 코드 작성 과정입니다.

>의사 코드는 컴퓨터 프로그램이나 알고리즘이 수행해야 할 내용을 우리가 사용하는 언어(한국어 또는 영어 등등)로 간략히 서술해 놓은
>것입니다. 따라서 우리는 의사 코드를 보고 작성된 프로그램 코드가 어떤 동작을 하는지 대략적으로 알 수 있게 됩니다.

그렇다면 왜 의사 코드를 사용할까요? 
1. 프로그램 코드를 작성하기 전, 프로그램의 전체적인 개요를 파악할 수 있어서 개발자에게 큰 도움이 됩니다.
2. 프로그램 코드를 직접 테스트해서 디버깅하기 전, 의사 코드를 통해 미리 설계 단계 속 오류를 찾아낼 수 있습니다.
3. 의사 코드 자체가 실제 코드의 주석(Comment)가 되기 때문에, 주석(Comment)에 대한 부담을 덜 수 있습니다.

[참고한 블로그](https://sujinlee.me/pseudocode/)

이처럼 의사 코드는 프로그래밍 과정에서 빠질 수 없는 설계 단계 중 하나 입니다. 하지만 아마도 대부분의 사람들은 프로그래밍 언어를 배우면서 처음부터 의사 코드 작성 방법에 대해 공부하지 않았을 것입니다. 필자 또한 그렇습니다...ㅎㅎ. 아무튼! 이 글을 읽고 계는 여러분들은 아마도 의사 코드 작성에 많이 애쓰고 있을 거라고 생각합니다.

그래서 제가 의사 코드 작성에 도움이 될 수 있도록 가이드라인은 소개해드리려고 합니다. 단, 이 가이드라인은 정답이 아닌 의사 코드 작성에 대한 방법, 틀, 예시이기 때문엔 무조건적으로 따를 필요 없이 여러분들에게 도움이 되는 것만 골라서 사용하시면 됩니다!

그리고 이 의사 코드 가이드라인은 예제를 출제하기 위한 것이 목적이라는 걸 확인해주시면 감사하겠습니다~

## 의사 코드 작성 전에 할 수 있는 것

- __프로그램 목적을 작성한다.__
```
ex) C++ 표준 라이브러리의 string 클래스와 비슷한 동작을 하는 String 클래스를 구현한 프로그램입니다.
```

- __구현할 주요 내용들을 작성한다.__
```
C++ 표준 라이브러리의 string 클래스와 비슷한 동작을 하는 String 클래스를 구현한다고 가정했을 때,
ex) "디폴트 생성자, 사용자 정의 생성자, 복사 생성자, 복사 할당 연산자, 소멸자를 구현합니다."
```

- __예제 출제자는 문제의 난이도를 고려하여, 너무 어렵거나 예제 해결에 큰 도움을 주지 않는 code에 한해서
참고사항에 code와 사용해야 하는 이유를 함께 작성한다.__
```
ex) *** String 클래스는 string과는 다르게 std::cout으로 출력할 수 없기 때문에 아래 code를 사용하여 출력할 수 있도록 합니다. ***

friend std::ostream& operator<<( std::ostream& os, const String& str ); - 클래스 내 public에 선언

std::ostream& operator<<( std::ostream& os, const String& str ) - 클래스 외부에 정의
{
	os << str._str;
	return os;
}
```

#### 또한 다른 사람이 여러분들의 의사 코드의 전체적인 개요를 미리 파악할 수 있게 되며, 낯선 코드에 대한 거부감을 해소시킬 수 있게 됩니다.
---

## 의사 코드 작성
의사 코드 작성법에는 정확한 규칙이 존재하지 않습니다. 하지만 사람들 사이에서 통용되는 표현들과 본인의 입맛대로 사용하는 규칙들, 혹은 본인만의 새로운 규칙들을 만들어 의사 코드를 작성하곤 합니다. 

굳이 많은 표현과 규칙들을 사용한다고 해서 좋은 의사 코드가 되는 것도 아니고, 표현과 규칙들을 많이 사용하지 않고 한국어, 영어 등, 글로만 작성한다고 해서 좋은 의사 코드가 되는 것이 아닙니다. 

저는 찾아보거나 직접 만들어본 의사 코드 작성법 중 여러분들이 쉽게 의사 코드에 접근할 수 있도록 도와줄만한 표현, 규칙들을 설명해드리려고 합니다. 만약 더 알아보고 싶으신 분들은 구글에 검색해보거나 참조 라인에 있는 제가 참고했던 블로그들을 보시는 것을 추천드립니다.

- __의사 코드는 어떠한 프로그래밍 언어에도 의존하면 안된다.__
  - 의사 코드는 다른 사람이 볼 가능성도 있고, 프로그램 코드의 설계 단계이기 때문에, 특정한 프로그래밍 언어를 모른다고 그 의사 코드를 이해할 수 없게 만드는 것은 안된다고 생각합니다.
- __함수, 변수 네이밍 컨벤션( Naming Convention ) 정하기__
```
네이밍 컨벤션에
1. snake case ex) snake_case
2. camel case ex) camelCase
3. pascal case ex) PascalCase
```
- __한 줄에 한 문장을 작성한다.__ 
  - 한 문장 안에 들어가는 동사 개수는 1 or 2개로 제한. 
  ( 한 문장의 길이가 너무 길어지는 것을 대비해 동사의 개수를 제한해 보았습니다. 선택 사항입니다. )
- __들여쓰기를 사용한다.__
  - 들여쓰기는 의사 코드의 가독성을 향상시켜줍니다.
  - 코드 블록을 중괄호를 사용하여 나누는 것은 비추천합니다. 왜냐하면 프로그래밍 언어스러워 질 거 같기 때문입니다.
- __함수 혹은 조건문, 반복문이 발생하거나 특정한 표현을 사용했을 경우, ```콜론( : )``` 으로 표시하여 그 시작을 알린다.__
( 뜬금없이 왜 이런 기호를 쓰지 싶으시겠지만, 저의 글을 읽다 보면 이해가 되실 겁니다. 조금만 참아주세요..! )
- __함수에 대한 표현__
  ![image](https://user-images.githubusercontent.com/75879925/215533462-fac1efae-b603-417c-8736-8ea608bdd5a0.png)
  - 위 사진은 함수의 의사 코드 표현입니다. 여러분들은 이 표현을 그대로 사용하셔도 좋고, 변형시키셔도 좋고, 참고하여 새로운 표현을 만드셔도 좋습니다.
딱 예시만 올려놓으면 이해가 제대로 안될 것 같으니, 제가 위의 표현을 참고하여 새롭게 만든 함수 표현을 보여드리겠습니다.
  - 하지만 그 전에 알아두어야 하는 기호와 표현이 있습니다. 
![image](https://user-images.githubusercontent.com/75879925/216064522-89ee431d-2f54-4fb8-a07c-2e1b0ca92ba3.png)
첫 번째는 ```<-```인데요. 여러분들은 이 녀석이 ```"대입을 의미하는구나"```라고 이해하시면 됩니다. 그러면 질문 하나가 생기실 겁니다. 바로 "대입이면 변수에 쓰이는 기호 아닌가?"라는 질문입니다. 하지만 제가 이에 대한 규칙을 만들었으니 잠시 뒤에 보시죠!
두 번째는 ```입력, 출력```입니다. ```입력```은 함수의 인자로 넘어오는 것들에 대한 표현이고, ```출력```은 함수의 반환값에 대한 표현입니다.
이제 직접 눈으로 보시면 이해가 더 빠를 겁니다~
```
1. "<-" 사용. [규칙: 변수와 사용되면 대입, 함수와 사용되면 정의 혹은 호출을 의미한다.]
ex)
num1 <- 10
num2 <- 20
add <- ( num1, num2 ):
    return sum

":"이 붙어있으면 정의, 그렇지 않으면 호출을 의미한다고 생각하시면 됩니다.
add <- ( num1, num2 ):  ------- 정의
add <- ( num3, num4 ) --------- 호출

return을 사용하여 반환값을 나타낸 이유는 영어 단어로써 보았을 때 "돌려주다"라는 의미를 가지고 있어서
프로그래밍 언어를 모르는 사람도 이해할 수 있을 거라 생각했기 때문입니다.


2. "입력, 출력" 표현 사용.
ex)
num1 <- 10
num2 <- 20
입력: num1, num2
add:
    출력: sum

1번의 설명을 토대로 2번이 어떻게 사용될지 여러분들이 직접 유추해 보는 것이 도움이 될 겁니다~
```

- __조건문에 대한 표현__
  - 조건문을 나타낼 때는 ```if 혹은 IF```를 사용합니다.
  - 통용되는 표현이기도 하고, 영어의 뜻만 보았을 때, 누구나 조건의 의미를 떠올릴 수 있기 때문에 사용하기 좋다고 생각합니다.
```
IF new_left <= new_right:
	배열의 new_left번째의 값과 new_right번째의 값을 교환한다.
	new_left값을 하나 증가시킨다.
	new_right값을 하나 감소시킨다.

조건문이 어떻게 쓰이는지만 보시면 됩니다~
```
- __반목문에 대한 표현__
  - 반복문에는 for, while, do while, range based for( for each로 표현됨. )( 범위 기반 반복문 )가 있습니다. 대부분의 의사 코드에서는 반복문에 대한 표현으로 ```while, do while, range based for 반복문```을 자주 사용하는 것으로 보입니다. 이 표현 그대로 사용하셔도 됩니다. 하지만 프로그래밍 언어를 그대로 사용하기 싫으시다면 새로운 표현을 만드셔도 좋습니다!!
  - __while 표현식__ 형식
  - __do while 표현식__ 형식
  - __for each A in( on, etc.. ) B, end for__ 형식
  - __repeat until__ 형식
  - 여러 개의 예시가 있습니다. 여러분들은 각각의 표현들이 어떻게 쓰였는지만 빠르게 파악하시면 됩니다~
![image](https://user-images.githubusercontent.com/75879925/217298196-927ff81c-8612-49fc-a90c-0e3436f2c81a.png)
![image](https://user-images.githubusercontent.com/75879925/217298290-118accc7-e614-421d-9cbd-730a438a17a2.png)
![image](https://user-images.githubusercontent.com/75879925/217298332-343a0980-4fb9-4c99-91f6-b10e8e95f0a9.png)
![image](https://user-images.githubusercontent.com/75879925/217299239-f0ad77cf-f44d-4ded-b2b7-96cf077baeec.png)
![image](https://user-images.githubusercontent.com/75879925/217299273-55ca9e3d-5ea2-462c-b6ff-9f51dcc552a2.png)

- __연산자의 사용이 가능하다.__
  - 비교 연산으로 사용할 수 있는 기호들
    - ```같다, 같지 않다, <, >, <=, >=```
    - "같다, 같지 않다"에 대한 표현은 여러분들이 새로 만드셔도 됩니다.
  - 사칙연산으로 사용할 수 있는 기호들
    - ```+, -, /, *```
    - 나머지 연산자가 필요하다면 사용하셔도 됩니다.
  - 논리 연산으로 사용할 수 있는 표현들
    - ```and, or```
    - &와 |와 같은 기호를 사용하는 것보다 나을 것 같습니다.
    즉, 프로그래밍 언어를 모르는 사람이 봤을 때, 이해하기 더 쉬울 것 같다고 생각합니다.
- __수의 범위를 나타내고 싶을 땐, ```[, ], (, )``` 을 사용한다.__
```
ex)
- [1, 10] : 1 이상 10 이하
- [1, 10) : 1 이상 10 미만
- (1, 10] : 1 초과 10 이하
- (1, 10) : 1 초과 10 미만
```
- __선행 조건은 본격적인 의사 코드 위에 작성한다.__
  - 의사 코드 가장 윗부분 글의 순서는 프로그램 목적, 구현할 주요 내용, 선행 조건, 참고 코드로 하면 좋을 것 같습니다.
- __후행 조건은 그 조건이 발생하는 지점에 작성한다.__

## 의사 코드 작성 후에 할 수 있는 것
반환값이 있는 함수가 잘못된 값을 반환하고 있지는 않는지, 문맥을 검토하여 생략이 필요한 부분이 있는지 정도가 있습니다.

## 의사 코드 예시
```
//---------------------------------------------------------------------------------------------------
// QuickSort를 구현해보는 예제입니다.
//---------------------------------------------------------------------------------------------------
// pivot은 중간값으로 설정합니다.
//---------------------------------------------------------------------------------------------------
QuickSort <- ( list, left, right ):
	new_left <- left
	new_right <- right
	pivot을 설정한다.

	REPEAT UNTIL new_left <= new_right:
		배열의 new_left번째의 값이 pivot보다 크거나 같다면, new_left 값을 고정한다.
		그렇지 않다면 new_left값을 하나 증가시킨다.
		이 과정을 반복한다.

		배열의 new_right번째의 값이 pivot보다 작거나 같다면, new_right 값을 고정한다.
		그렇지 않다면 new_right값을 하나 감소시킨다.
		이 과정을 반복한다.

		IF new_left <= new_right:
			배열의 new_left번째의 값과 new_right번째의 값을 교환한다.
			new_left값을 하나 증가시킨다.
			new_right값을 하나 감소시킨다.

	IF left < new_right:
		QuickSort <- ( list, [left, new_right] )

	IF new_left < right:
		QuickSort <- ( list, [new_left, right] )
//---------------------------------------------------------------------------------------------------

#include <iostream>

void QuickSort( int array[ ], int L, int R );

int main( )
{
	int array[ 10 ]{ 5, 1, 4, 10, 3, 7, 2, 8, 9, 6 };

	QuickSort( array, 0, 9 );

	for ( const auto& num : array )
		std::cout << num << ' ';
	std::cout << std::endl;
}

void QuickSort( int array[ ], int L, int R )
{
	int left{ L }, right{ R };
	int pivot{ array[ ( L + R ) / 2 ] };

	while ( left <= right )
	{
		while ( array[ left ] < pivot )
			++left;
		
		while ( array[ right ] > pivot )
			--right;

		if ( left <= right )
		{
			int temp{ array[ left ] };
			array[ left ] = array[ right ];
			array[ right ] = temp;

			++left;
			--right;
		}
	}

	if ( L < right )
		QuickSort( array, L, right );

	if ( left < R )
		QuickSort( array, left, R );
}
```

## 참조
https://gist.github.com/camilstaps/59c4574ab8131fb83612a446606cbcba
https://builtin.com/data-science/pseudocode
https://brunch.co.kr/@markism/31
https://red-mimmu.tistory.com/m/2
https://www.geeksforgeeks.org/how-to-write-a-pseudo-code/
https://www.geeksforgeeks.org/what-is-pseudocode-a-complete-tutorial/
https://en.wikipedia.org/wiki/Pseudocode
https://en.wikiversity.org/wiki/Pseudocode
https://dev.to/flippedcoding/how-to-write-pseudo-code-2jfe
https://sparkeasy.in/2022/08/what-is-pseudocode.html

**Comment by Woon-2** [comment]
자료들을 좀 찾아봤습니다.
디스코드에서 설명해드리죵.

https://red-mimmu.tistory.com/2
https://sujinlee.me/pseudocode/
https://brunch.co.kr/@markism/31
https://www.geeksforgeeks.org/how-to-write-a-pseudo-code/
https://en.wikipedia.org/wiki/Pseudocode
https://app.code2flow.com/
https://marketplace.visualstudio.com/items?itemName=willumz.generic-pseudocode
http://flowgorithm.org/about/screenshots.html
https://builtin.com/data-science/pseudocode
https://gist.github.com/camilstaps/59c4574ab8131fb83612a446606cbcba

검색했던 키워드는   
수도 코드, 수도 코드 작성법, 의사 코드, 좋은 의사 코드, pseudo code, pseudo code writing, pseudo code editor, good pseudo code
등등입니다.

구글과 스택오버플로우, 아마존(책을 찾아 목차에서 키워드들을 발견해내기 위한 용도로)에서 검색했습니다.   
깃허브에서도 검색하려 했는데 시간이 모자라네요.

**Comment by Woon-2** [comment]
![다운로드](https://user-images.githubusercontent.com/73771162/213202739-83e30693-0794-4458-9321-f6569df5577a.jpg)
![다운로드 (1)](https://user-images.githubusercontent.com/73771162/213202750-5a5e6c68-ae24-4dae-a2f1-0c7770b7580f.jpg)

이산수학이라는 키워드가 머릿속에서 번뜩이더군요.    

제가 수학이라는 학문에 대해 갖고 있는 인식 중 하나는,   
수학이 일상적인 것에 대한 표현들을 연구한다는 것입니다.
일반화, 추상화, 캡슐화 등등의 개념들이 암묵적으로 수학의 특성과 연관되어 있죠.

이산수학에서 그래프의 표현을 다루었고,   
그래프는 알고리즘의 주요 도구니   
보통 알고리즘을 표현하는데 쓰이는 수도코드에   
이산수학이 먹혀들 것이라 판단했습니다.

그래서 처음에 "이산수학 알고리즘", "이산수학 수도코드"의 키워드로 검색했으나    
유용한 정보는 찾지 못했습니다.   

그러다가 컴퓨터과학 논문들에는 반드시 수도코드가 포함되어 있을 것이고      
논문에 게재될 정도라면 반드시 잘 짜여진 수도코드일 것이라고 판단하여,   
"논문 수도코드"라는 키워드로 검색했습니다.

그 결과 위 자료를 얻을 수 있었습니다.   
오일석, 『기계 학습』, (한빛아카데미, 2017)   
을 읽으면서 보았던 수도코드와도 생김새가 유사했습니다.

내용도 내용이지만,   
제가 이런 자료 조사 과정을 거쳤다는 걸 보여드리는 것도   
좋은 영양분이 될 것 같습니다.

활용까지는 시간이 걸리더라도,    
한번 맛봐보시죠.
